spark/src/test/scala/org/apache/spark/sql/delta/managedcommit/ManagedCommitSuite.scala

```diff
@@ -116,6 +123,405 @@ class ManagedCommitSuite
+
+        // Step-5: Read the table again and assert that the right APIs are used
+        resetMetrics()
+        assert(
+          sql(s"SELECT * FROM delta.`$tablePath`").collect().toSet === (0 to 4).map(Row(_)).toSet)
+        // since this was hot query, so no new snapshot was created as part of this
+        // deltaLog.update() and so commit-store is not initialized again.
+        assert((builder1.numBuildCalled, builder2.numBuildCalled) === (0, 0))
+        // Since this is dataframe read, so we invoke deltaLog.update() twice and so GetCommits API
+        // is called twice.
+        assert((cs1.numGetCommitsCalled, cs2.numGetCommitsCalled) === (0, 2))
+
+        // Step-6: Clear cache and simulate cold read again.
+        // We will firstly create snapshot from 0.json. Then contact CS1 and update snapshot to v2.
+        // Then we contact CS2 again to recursively update snapshot and find that v3, v4 exist.
+        // Then we recursively update snapshot to v4 and see no change in commit owner and exit
+        // recursion.
+        DeltaLog.clearCache()
+        resetMetrics()
+        assert(
+          sql(s"SELECT * FROM delta.`$tablePath`").collect().toSet === (0 to 4).map(Row(_)).toSet)
+        assert((cs1.numGetCommitsCalled, cs2.numGetCommitsCalled) === (1, 2))
+        assert((builder1.numBuildCalled, builder2.numBuildCalled) === (1, 2))
+      }
+    }
+  }
+
+  // This test has the following setup:
+  // Setup: There are two commit-stores: cs1 and cs2.
+  //  Any commit recorded in CS1 is also recorded in CS2.
+  //  Any commit recorded in CS2 is not recorded in CS1.
+  // 1. Table is created with CS1 as commit-owner. So 0th commit is persisted in both CS1 and CS2.
+  // 2. Do another commit (v1) on table. Again this will be recorded in both CS1 and CS2.
+  // 3. Take a reference to current DeltaLog and clear the cache. This deltaLog object currently
+  //    points to the latest table snapshot i.e. v1.
+  // 4. Do commit v2 on the table.  Again this will be recorded in both CS1 and CS2.
+  // 5. Do commit v3 on table. As part of this, change commit-owner to CS2. This will be recorded
+  //    in both commit stores CS1, CS2 as CS1 was managing the table when the commit happened.
+  // 6. Do commit v4 on table. This will be recorded in CS2 only as CS2 is the owner now.
+  // 7. Invoke deltaLog.update() on the old deltaLog object which is still pointing to v1.
+  //    - While doing this, we will inject failure in CS2 so that it fails twice when cs2.getCommits
+  //      is called.
+  //    - Because of this old delta log will firstly contact cs1 and get newer commits  i.e. v2/v3
+  //      and create a snapshot out of it. Then it will contact cs2 and fail. So deltaLog.update()
+  //      won't succeed and throw exception. But underlying DeltaLog object now has reference to v3.
+  //      The recorded timestamp for this must be commit timestamp and not clock timestamp.
```

Comment:
How could `DeltaLog` instance have a reference to v3, when `getUpdatedSnapshot` only installs the new snapshot until after the commit-owner-changed loop exits?

Reply:
The description here is out-of-sync from the actual logic. Fixing this.